app/src/lib/stores/helpers/ahead-behind-updater.ts

```diff
@@ -70,38 +49,33 @@ export class AheadBehindUpdater {
     from: string,
     to: string
   ): Promise<IAheadBehind | null> {
-    return new Promise((resolve, reject) => {
-      if (this.comparisonCache.has(from, to)) {
-        resolve(this.comparisonCache.get(from, to))
-        return
-      }
+    if (this.comparisonCache.has(from, to)) {
+      return this.comparisonCache.get(from, to)
+    }
 
-      this.executeTask(from, to, (error, result) =>
-        error !== null ? reject(error) : resolve(result)
-      )
-    })
+    return this.executeTask(from, to)
   }
 
-  private executeTask = (
+  private executeTask = async (
     from: string,
-    to: string,
-    callback: (error: Error | null, result: IAheadBehind | null) => void
-  ) => {
+    to: string
+  ): Promise<IAheadBehind | null> => {
     if (this.comparisonCache.has(from, to)) {
-      return
+      return null
```

Comment:
This seems really weird (and I know it's not your fault). Shouldn't this be `return this.comparisonCache.get(from, to)` ?

```suggestion
      return this.comparisonCache.get(from, to)
```

And once we've done that is there even a need for the `executeTask` method or could we just roll up the logic into `executeAsyncTask`.

The previous logic seems super buggy, why would we just skip calling the callback?

Reply:
Yes, I think returning the value of the `comparisonCache` it's much better. I thought it didn't matter much since the queue results are not directly used: there's only a not-null check on the `success` callback but if I understand the logic correctly, it won't matter if we return `null`.

> The previous logic seems super buggy, why would we just skip calling the callback?

Uhm good point! that seems like a potential serious issue 😅 

The whole logic of this module is quite weird... what I think makes it hard to reason is that the queue is only used to populate the cache object which then gets exported outside of this module.

In one of my first PRs in Desktop I had to deal with an issue with this approach and I introduced the `executeAsync` method to be able to access the aheadBehind information of a single `from/`to` combination without having to execute the whole queue (see https://github.com/desktop/desktop/pull/9319).